satellite/gracefulexit/endpoint.go

```diff
@@ -862,3 +686,216 @@ func (endpoint *Endpoint) updatePointer(ctx context.Context, originalPointer *pb
+		exitStatusRequest.ExitSuccess = false
+	} else {
+		exitStatusRequest.ExitSuccess = true
+	}
+
+	if exitStatusRequest.ExitSuccess {
+		mon.Meter("graceful_exit_success").Mark(1)
+	} else {
+		mon.Meter("graceful_exit_fail_max_failures_percentage").Mark(1)
+	}
+
+	return exitStatusRequest, nil
+
+}
+
+func (endpoint *Endpoint) validatePointer(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem) (*pb.RemotePiece, error) {
+	remote := pointer.GetRemote()
+	nodeID := incomplete.NodeID
+
+	pieces := remote.GetRemotePieces()
+	var nodePiece *pb.RemotePiece
+	for _, piece := range pieces {
+		if piece.NodeId == nodeID && piece.PieceNum == incomplete.PieceNum {
+			nodePiece = piece
+		}
+	}
+
+	if nodePiece == nil {
+		endpoint.log.Debug("piece no longer held by node", zap.Stringer("node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+
+		err := endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+
+	}
+	return nodePiece, nil
+}
+
+func (endpoint *Endpoint) generateFindNodesRequest(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem) (*overlay.FindStorageNodesRequest, error) {
+	// validate pointer state
+	nodePiece, err := endpoint.validatePointer(ctx, pointer, incomplete)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	nodeID := incomplete.NodeID
+	pieces := pointer.GetRemote().GetRemotePieces()
+	if len(pieces) > redundancy.OptimalThreshold() {
+		endpoint.log.Debug("pointer has more pieces than required. removing node from pointer.", zap.Stringer("node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+
+		_, err = endpoint.metainfo.UpdatePieces(ctx, string(incomplete.Path), pointer, nil, []*pb.RemotePiece{nodePiece})
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+
+		err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+
+		return nil, nil
+	}
+
+	// calculate piece size
+	pieceSize := eestream.CalcPieceSize(pointer.GetSegmentSize(), redundancy)
+
+	// populate excluded node IDs
+	excludedNodeIDs := make([]storj.NodeID, len(pieces))
+	for i, piece := range pieces {
+		excludedNodeIDs[i] = piece.NodeId
+	}
+
+	return &overlay.FindStorageNodesRequest{
+		RequestedCount: 1,
+		FreeBandwidth:  pieceSize,
+		FreeDisk:       pieceSize,
+		ExcludedNodes:  excludedNodeIDs,
+	}, nil
+}
+
+func (endpoint *Endpoint) verifyPieceTransferred(ctx context.Context, message *pb.StorageNodeMessage_Succeeded, transfer *pendingTransfer) error {
```

Comment:
Maybe we can move this and `validatePointer` (and any other similar functions) to a file called `validation.go`. Similar to the metainfo package.